exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/impl/SqlPatternContainsMatcher.java

```diff
@@ -19,44 +19,283 @@
+    }
   }
 
   @Override
   public int match(int start, int end, DrillBuf drillBuf) {
+    return matcherFcn.match(start, end, drillBuf);
+  }
+
+  //--------------------------------------------------------------------------
+  // Inner Data Structure
+  // --------------------------------------------------------------------------
+
+  /** Abstract matcher class to allow us pick the most efficient implementation */
+  private abstract class MatcherFcn {
+    protected final byte[] patternArray;
+
+    protected MatcherFcn() {
+      assert patternByteBuffer.hasArray();
+
+      patternArray = patternByteBuffer.array();
+    }
+
+    /**
+     * @return 1 if the pattern was matched; 0 otherwise
+     */
+    protected abstract int match(int start, int end, DrillBuf drillBuf);
+  }
+
+  /** Handles patterns with length one */
+  private final class Matcher1 extends MatcherFcn {
 
-    if (patternLength == 0) { // Everything should match for null pattern string
-      return 1;
+    private Matcher1() {
+      super();
     }
 
-    final int txtLength = end - start;
+    /** {@inheritDoc} */
+    @Override
+    protected final int match(int start, int end, DrillBuf drillBuf) {
+      final int lengthToProcess = end - start;
+      final byte firstPattByte  = patternArray[0];
 
-    // no match if input string length is less than pattern length
-    if (txtLength < patternLength) {
+      // simplePattern string has meta characters i.e % and _ and escape characters removed.
+      // so, we can just directly compare.
+      for (int idx = 0; idx < lengthToProcess; idx++) {
+        byte inputByte = drillBuf.getByte(start + idx);
+
+        if (firstPattByte != inputByte) {
+          continue;
+        }
+        return 1;
+      }
       return 0;
     }
+  }
 
+  /** Handles patterns with length two */
+  private final class Matcher2 extends MatcherFcn {
 
-    final int outerEnd = txtLength - patternLength;
+    private Matcher2() {
+      super();
+    }
 
-    outer:
-    for (int txtIndex = 0; txtIndex <= outerEnd; txtIndex++) {
+    /** {@inheritDoc} */
+    @Override
+    protected final int match(int start, int end, DrillBuf drillBuf) {
+      final int lengthToProcess = end - start - 1;
+      final byte firstPattByte  = patternArray[0];
+      final byte secondPattByte = patternArray[1];
 
       // simplePattern string has meta characters i.e % and _ and escape characters removed.
       // so, we can just directly compare.
-      for (int patternIndex = 0; patternIndex < patternLength; patternIndex++) {
-        if (patternByteBuffer.get(patternIndex) != drillBuf.getByte(start + txtIndex + patternIndex)) {
-          continue outer;
+      for (int idx = 0; idx < lengthToProcess; idx++) {
+        final byte firstInByte = drillBuf.getByte(start + idx);
+
+        if (firstPattByte != firstInByte) {
+          continue;
+        } else {
+          final byte secondInByte = drillBuf.getByte(start + idx +1);
```

Comment:
same comment as above. No need to save in local variables

Reply:
same answer.